app/grandchallenge/cases/image_builders/dicom.py

```diff
@@ -167,49 +170,56 @@ def _process_dicom_file(dicom_ds):  # noqa: C901
     content_times = []
     exposures = []
 
+    origin = None
+    origin_diff = np.array((0, 0, 0), dtype=float)
+    n_origins = 0
+    for partial in dicom_ds.headers:
+        ds = partial["data"]
+        if "ImagePositionPatient" in ds:
+            file_origin = np.array(ds.ImagePositionPatient, dtype=float)
+            if origin is not None:
+                diff = file_origin - origin
+                origin_diff = origin_diff + diff
+            origin = file_origin
+            n_origins += 1
+    avg_origin_diff = tuple(origin_diff / n_origins)
+    z_i = avg_origin_diff[-1]
     for index, partial in enumerate(dicom_ds.headers):
         ds = pydicom.dcmread(str(partial["file"]))
+
         # Apply RescaleSlope and RescaleIntercept
         pixel_array = float(
             getattr(ds, "RescaleSlope", 1)
         ) * ds.pixel_array + float(getattr(ds, "RescaleIntercept", 0))
         if len(ds.pixel_array.shape) == dimensions:
             dcm_array = pixel_array
             break
+        z_index = index if z_i >= 0 else len(dicom_ds.headers) - index
```

Comment:
I think this should be len(dicom_ds.headers) - index - 1, because index is 0-based, right? otherwise, the last slice has index len(dicom_ds.headers) which 1 too large

Reply:
Also, use 0.0 instead of 0 to make clear that you are doing a float comparison